postprocessing-emissions/src/main/java/org/matsim/viz/postprocessing/emissions/DataGenerator.java

```diff
@@ -46,12 +47,27 @@ public void generate(Input<Visualization> input) {
                 .build();
 
         log.info("Start processing emissions. This may take a while.");
-        val json = analyzer.processToJsonString(input.getInputFiles().get(EVENTS_KEY).getPath().toString());
 
-        log.info("Finished processing emissions. Write result to database");
+        analyzer.processTimeBinsWithEmissions(input.getInputFiles().get(EVENTS_KEY).getPath().toString());
 
+        // all of the bins need to be successful, or the transaction should fail.
         session.beginTransaction();
-        mergedViz.setData(json);
+
+        Tuple<Double, String> result = analyzer.processNextTimeBin();
+
+        while (result != null) {
+            log.info("-- JSON size: " + result.getSecond().length());
+
+            Bin b = new Bin();
+            b.setStartTime(result.getFirst());
+            b.setData(result.getSecond());
+            mergedViz.addBin(b);
+
+            result = analyzer.processNextTimeBin();
```

Comment:
Hm, we can leave it like this for now. But having a while loop with a null condition looks error prone. Maybe the Iterator Interface is what you want here? I'm not sure, but maybe that would come close to the yield return we were talking about.

Reply:
This is the main thing I was wondering about in this commit. It feels dirty to me as well. In Java I'm not sure what the least-errorprone / most-appropriate form would be. I'll look at the Iterator interface.